src/p2p/swarm.rs

```diff
@@ -310,71 +324,51 @@ impl NetworkBehaviour for SwarmApi {
-        // surprising happens.
-        for failed in self.pending_connections.remove(peer_id).unwrap_or_default() {
-            self.connect_registry
-                .finish_subscription(failed.into(), Err("addresses exhausted".into()));
-        }
-    }
-
-    fn inject_addr_reach_failure(
+    fn inject_dial_failure(
         &mut self,
-        peer_id: Option<&PeerId>,
-        addr: &Multiaddr,
-        error: &dyn std::error::Error,
+        peer_id: Option<PeerId>,
+        _handler: Self::ConnectionHandler,
+        error: &DialError,
     ) {
-        trace!("inject_addr_reach_failure {} {}", addr, error);
-
         if let Some(peer_id) = peer_id {
-            match self.pending_connections.entry(*peer_id) {
+            match self.pending_connections.entry(peer_id) {
                 Entry::Occupied(mut oe) => {
                     let addresses = oe.get_mut();
-                    let addr = MultiaddrWithPeerId::try_from(addr.clone())
-                        .expect("dialed address contains peerid in libp2p 0.38");
-                    let pos = addresses.iter().position(|a| *a == addr);
 
-                    if let Some(pos) = pos {
-                        addresses.swap_remove(pos);
-                        self.connect_registry
-                            .finish_subscription(addr.into(), Err(error.to_string()));
+                    match error {
+                        DialError::Transport(multiaddrs) => {
+                            for (addr, error) in multiaddrs {
+                                let addr = MultiaddrWithPeerId::try_from(addr.clone())
+                                    .expect("to recieve an MultiAddrWithPeerId from DialError");
+                                self.connect_registry
+                                    .finish_subscription(addr.into(), Err(error.to_string()));
+                            }
+
+                            let peer_ids = multiaddrs
+                                .into_iter()
+                                .map(|(addr, _err)| {
+                                    MultiaddrWithPeerId::try_from(addr.clone()).unwrap()
+                                })
+                                .collect::<Vec<_>>();
+
+                            addresses.retain(|peer_id| peer_ids.iter().any(|id| peer_id == id));
+                        }
+                        DialError::WrongPeerId {
+                            obtained: _,
+                            endpoint: _,
+                        } => {
+                            for addr in addresses.iter() {
+                                self.connect_registry.finish_subscription(
+                                    addr.clone().into(),
+                                    Err(error.to_string()),
+                                );
+                            }
+
+                            addresses.clear();
+                        }
+                        err => trace!("unhandled DialError {}", err),
```

Comment:
Should we trace every DialError? Was thinking about the WrongPeerId failure above and I'm not sure if it would otherwise be reported to the user.

Reply:
I would probably go with

```suggestion
                        error => {
                          warn!(?error, "unexpected DialError; some futures might never complete");               
                        },
```

Unless you can see a way to finish all of the subscriptions under this peerid? I cannot at least, we had not considered this situation originally as mentioned the events used to all arrive.